proposals/texture-buffers.md

```diff
@@ -0,0 +1,405 @@
+
+
+# Appendix A: Implementation details
+
+
+### Vulkan
+
+In Vulkan, a `read_write` texture buffer would map to a storage texel buffer decorated as `Coherent`, and shader accesses would be performed with the `OpImageRead` and `OpImageWrite` instructions.
+A texture buffer with a `read-only` access mode could use a uniform texel buffer instead, which would use `OpImageFetch` instead `OpImageRead`.
+
+When the `TEXTURE_BUFFER` usage flag is set on buffer creation, both of the Vulkan texel buffer bits would be set:<br/>
+`VK_BUFFER_USAGE_UNIFORM_TEXEL_BUFFER_BIT | VK_BUFFER_USAGE_STORAGE_TEXEL_BUFFER_BIT`
+
+The [required image formats](https://registry.khronos.org/vulkan/specs/1.2-extensions/html/vkspec.html#features-required-format-support) for storage texel buffers includes:
+
+```
+R8G8B8A8_UNORM
+R8G8B8A8_UINT
+R8G8B8A8_SINT
+R16G16B16A16_UINT
+R16G16B16A16_SINT
+R16G16B16A16_SFLOAT
+R32_UINT
+R32_SINT
+R32_SFLOAT
+R32G32_UINT
+R32G32_SINT
+R32G32_SFLOAT
+R32G32B32A32_UINT
+R32G32B32A32_SINT
+R32G32B32A32_SFLOAT
+```
+
+For the other formats, [gpuinfo.org](https://vulkan.gpuinfo.org/listbufferformats.php) has information on how widespread support is.
+For 1- and 2-channel `R{8,16}_{SINT,UINT,SFLOAT}`, support is currently around 80% for storage texel buffers.
+
+**TODO:** We should do this query against WebGPU's baseline requirements, as the percentage for devices we actually support may be higher.
+
+Vulkan has a `maxTexelBufferElements` limit for the maximum size of a texel buffer.
+[gpuinfo.org shows that](https://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxTexelBufferElements&platform=all) more than 85% of devices support 128MB texel buffers.
+
+
+### Metal
+
+Metal has a `texture_buffer` type that provides similar functionality, which was introduced in Metal 2.1.
+The [Metal Feature Set Tables](https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf) show the supported formats for each access mode.
+Unnormalized integer and floating point formats are supported for all access modes, as is `RGBA8Unorm`.
+Using a `read-write` access mode also requires support for the Tier 2 `MTLReadWriteTextureTier`.
+A `mem_texture` fence would be needed to make texture buffer writes visible within an invocation.
+
+To get coverage on older Metal versions, it would be possible to polyfill by using a regular device buffer and doing the format conversions inside the shader.
+This requires that the storage format is specified inside the shader.
+
+The maximum texture buffer size is 64MB for the Apple2 GPU family, and 256MB for Apple3 and above.
+
+
+### D3D12
+
+In D3D12, a texture buffer can map to an Unordered Access View (UAV) for a buffer with a `DXGI_FORMAT`, and that UAV can be accessed in the shader with 32-bit result types.
+See [Typed unordered access view (UAV) loads](https://docs.microsoft.com/en-us/windows/win32/direct3d12/typed-unordered-access-view-loads).
+The `RWByteAddressBuffer` should be prefixed with `globallycoherent`.
```

Comment:
I think this should be `RWBuffer` instead. We also need to specify `unorm`/`snorm` if the format is normalized, as per:
https://learn.microsoft.com/en-us/windows/win32/direct3d12/typed-unordered-access-view-loads#using-unorm-and-snorm-typed-uav-loads-from-hlsl